compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -1384,422 +1630,180 @@ TR::Register *OMR::Power::TreeEvaluator::returnEvaluator(TR::Node *node, TR::Cod
+      {
+      TR::Node *tmpNode = trueNode;
+      trueNode = falseNode;
+      falseNode = tmpNode;
 
-static TR::InstOpCode::Mnemonic cmp2cmpi(TR::ILOpCodes op, TR::CodeGenerator *cg)
-    {
-    switch (op)
-       {
-       case TR::icmpeq:
-       case TR::icmpne:
-       case TR::icmplt:
-       case TR::icmpge:
-       case TR::icmpgt:
-       case TR::icmple:
-       case TR::bcmpeq:
-       case TR::bcmpne:
-       case TR::bcmplt:
-       case TR::bcmpge:
-       case TR::bcmpgt:
-       case TR::bcmple:
-          return TR::InstOpCode::cmpi4;
-       case TR::iucmplt:
-       case TR::iucmpge:
-       case TR::iucmpgt:
-       case TR::iucmple:
-       case TR::bucmplt:
-       case TR::bucmpge:
-       case TR::bucmpgt:
-       case TR::bucmple:
-          return TR::InstOpCode::cmpli4;
-       case TR::lcmpeq:
-       case TR::lcmpne:
-       case TR::lcmplt:
-       case TR::lcmpge:
-       case TR::lcmpgt:
-       case TR::lcmple:
-          return cg->comp()->target().is64Bit() ? TR::InstOpCode::cmpi8 : TR::InstOpCode::cmpi4;
-       case TR::lucmplt:
-       case TR::lucmpge:
-       case TR::lucmpgt:
-       case TR::lucmple:
-       case TR::acmpeq:
-       case TR::acmpne:
-       case TR::acmplt:
-       case TR::acmpge:
-       case TR::acmpgt:
-       case TR::acmple:
-          return cg->comp()->target().is64Bit() ? TR::InstOpCode::cmpli8 : TR::InstOpCode::cmpli4;
-       default:
-       TR_ASSERT(false, "assertion failure");
-       }
-    return TR::InstOpCode::bad;
-    }
+      return true;
+      }
 
+   return false;
+   }
 
+// Also handles bselect, sselect, aselect, and lselect
 TR::Register *OMR::Power::TreeEvaluator::iselectEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    {
-   TR::DataType type = node->getType();
-   bool two_reg = (cg->comp()->target().is32Bit()) && type.getDataType() == TR::Int64;
-   TR::Register *resultReg = two_reg ?
-                            cg->allocateRegisterPair(cg->allocateRegister(),cg->allocateRegister()) :
-                            (type.getDataType() == TR::Float ?
-                            cg->allocateSinglePrecisionRegister() :
-                            (type.getDataType() == TR::Double ?
-                            cg->allocateRegister(TR_FPR) :
-                            cg->allocateRegister(TR_GPR)));
-
-   // Mark the result reg as collected reference if required.
-   if (!node->isNotCollected())
-      {
-      resultReg->setContainsCollectedReference();
-      }
+   TR::Node *condNode = node->getFirstChild();
+   TR::Node *trueNode = node->getSecondChild();
+   TR::Node *falseNode = node->getThirdChild();
 
-   TR::InstOpCode::Mnemonic move_opcode = (type.isIntegral() || type.isAddress()) ? TR::InstOpCode::mr : TR::InstOpCode::fmr;
-   TR::Node * firstChild = node->getFirstChild();
+   TR::Register *trgReg;
+   TR::Register *condReg = cg->allocateRegister(TR_CCR);
 
-   if (firstChild->getOpCode().isBooleanCompare() &&
```

Comment:
I think isBooleanCompare() check was removed. I am slightly concerned that it can be some other expression, a constant for example.

Reply:
The removal of this check in the evaluator was intentional. Handling for this optimization has been moved into the new `evaluateToConditionRegister` function where it can be shared with other evaluators that can perform the same optimization, e.g. ZEROCHK. In the event that the node passed in is not a boolean compare, that function will simply evaluate the node to a GPR and compare the result against 0, so it will work regardless.